references/detection/train_pytorch_DDP.py

```diff
@@ -0,0 +1,546 @@
+    if rank==0 and args.wb:
+        run = wandb.init(
+            name=exp_name,
+            project="text-detection",
+            config={
+                "learning_rate": args.lr,
+                "epochs": args.epochs,
+                "weight_decay": args.weight_decay,
+                "batch_size": args.batch_size,
+                "architecture": args.arch,
+                "input_size": args.input_size,
+                "optimizer": "adam",
+                "framework": "pytorch",
+                "scheduler": args.sched,
+                "train_hash": train_hash,
+                "val_hash": val_hash,
+                "pretrained": args.pretrained,
+                "rotation": args.rotation,
+                "amp": args.amp,
+            },
+        )
+
+    # Create loss queue
+    min_loss = np.inf
+    if args.early_stop:
+        early_stopper = EarlyStopper(patience=args.early_stop_epochs, min_delta=args.early_stop_delta)
+
+    print('5. going inside training loop')
+    
+    # Training loop
+    for epoch in range(args.epochs):
+        fit_one_epoch(model, train_loader, batch_transforms, optimizer, scheduler, amp=args.amp)
+        
+        if rank == 0:
+            # Validation loop at the end of each epoch
+            val_loss, recall, precision, mean_iou = evaluate(model, val_loader, batch_transforms, val_metric,args, amp=args.amp)
+            if val_loss < min_loss:
+                print(f"Validation loss decreased {min_loss:.6} --> {val_loss:.6}: saving state...")
+                torch.save(model.module.state_dict(), f"./{exp_name}.pt")
+                min_loss = val_loss
+            if args.save_interval_epoch:
+                print(f"Saving state at epoch: {epoch + 1}")
+                torch.save(model.state_dict(), f"./{exp_name}_epoch{epoch + 1}.pt")
+            log_msg = f"Epoch {epoch + 1}/{args.epochs} - Validation loss: {val_loss:.6} "
+            if any(val is None for val in (recall, precision, mean_iou)):
+                log_msg += "(Undefined metric value, caused by empty GTs or predictions)"
+            else:
+                log_msg += f"(Recall: {recall:.2%} | Precision: {precision:.2%} | Mean IoU: {mean_iou:.2%})"
+            print(log_msg)
+            # W&B
+            if args.wb:
+                wandb.log({
+                    "val_loss": val_loss,
+                    "recall": recall,
+                    "precision": precision,
+                    "mean_iou": mean_iou,
+                })
+            if args.early_stop and early_stopper.early_stop(val_loss):
+                print("Training halted early due to reaching patience limit.")
+                break
+    print('6. coming out of training loop')       
+    if rank == 0:
+        if args.wb:
+            run.finish()
+
+        if args.push_to_hub:
+            push_to_hf_hub(model, exp_name, task="detection", run_config=args)
+    print('7. initializing destroy_process_group ')    
+    dist.destroy_process_group()
```

Comment:
It's the end of the "program" and should not be needed here :) -> remove

same for the print statement